### PR TITLE
Make the prefork mode more robust

### PR DESCRIPTION
**Pull request by panjf2000**
The main process will exit if one of the prefork child processes doesn't complete successfully under the current prefork mode, so it ought to make sure that all child processes run independently and the main process will only exit after all child processes are finished.

### CODE REVIEW

**Comment by savsgio** [comment]
HI @panjf2000,

It's not necessary to wait to the all child procs will finish, because if the master process dies, the children should die too.
Also, you are collected all error and just return the first one.

```go
return p.prefork(addr)[0]
```

Think that, if you wait to all child procs dies/end, you will lose performance because maybe you couldn't know it, but if master dies when any of child procs dies, you could know quickly that the server is dead and not lose performance due to the some cores doesn't work.

But we wait for @erikdubbelboer, to know their opinion.

**Comment by panjf2000** [comment]
> It's not necessary to wait to the all child procs will finish, because if the master process dies, the children should die too.

I know that, but this PR doesn't intend to make sure all child processes to exit after the main process exit, it try to avoid that any child process to break the main process.

> Also, you are collected all error and just return the first one.

Yes, you are right but the root cause of it is that the definition of `ListenAndServe` only return one single `error` so I have to pick only one error to return:
https://github.com/valyala/fasthttp/blob/695f713fcf594b33fd8c126cfa8dc9f119d4391b/server.go#L88
also, it printed those else errors to stdout, unfortunately it is only way for now.

> Think that, if you wait to all child procs dies/end, you will lose performance because maybe you couldn't know it, but if master dies when any of child procs dies, you could know quickly that the server is dead and not lose performance due to the some cores doesn't work.

You got a point but in my opinion, it makes more sense to choose robustness over performance in this case because fasthttp is fast enough, it can still provide high-performance services even though it does lose a little bit of performance because of some child prefork processes breaking down, so it shouldn't shut the entire server down just because one or two prefork child processes break down.

@savsgio

**Comment by savsgio** [comment]
What do you think @erikdubbelboer ?

**Comment by erikdubbelboer** [comment]
I think you're both wrong 😄 . When a parent process dies doesn't mean the children also automatically dies: https://unix.stackexchange.com/questions/158727/is-there-any-unix-variant-on-which-a-child-process-dies-with-its-parent

I think if one child dies it should be restarted. Up to a certain (configuratble) limit at which point `ListenAndServe` should return with an error. But before that, it should clean up/kill all the children. That way developers using this can handle failure in the way that they would like.

**Comment by panjf2000** [comment]
> When a parent process dies doesn't mean the children also automatically dies: https://unix.stackexchange.com/questions/158727/is-there-any-unix-variant-on-which-a-child-process-dies-with-its-parent

It is right in general, but we were talking about goroutines, you see, in the previous code the `cmd.Run()` was wrapped into a goroutine, so when the goroutine that starts other goroutines dies, other goroutines die, then the child processes started in goroutines become orphan processes and adopted by init process, which I think it is also a bad thing.

> I think if one child dies it should be restarted. Up to a certain (configuratble) limit at which point ListenAndServe should return with an error. But before that, it should clean up/kill all the children. That way developers using this can handle failure in the way that they would like.

So you are suggesting that I should add code that restarts child processes automatically when it finds any child process dies?

File: prefork/prefork.go [hunk 1]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
+				err = errors.New("exceeding the value of RecoverThreshold")
+				break
+			}
+
+			/* #nosec G204 */
+			cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			cmd.ExtraFiles = p.files
+			childProcs[cmd.Process.Pid] = cmd
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: Isn't `cmd.Process.Pid` here still the old value as `cmd.Run()` hasn't started yet?

File: prefork/prefork.go [hunk 2]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
+	}()
+
+	for i := 0; i < goMaxProcs; i++ {
 		/* #nosec G204 */
 		cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
 		cmd.ExtraFiles = p.files
+		if err = cmd.Start(); err != nil {
+			log.Printf("failed to start a child prefork process, error: %v\n", err)
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: I would prefer to not use the `log` package. This was users don't have much control over the output of their program. We don't do this anywhere else in fasthttp either. Better is doing something similar to https://godoc.org/github.com/valyala/fasthttp#Server.Logger

File: prefork/prefork.go [hunk 3]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
+
+	defer func() {
+		for _, proc := range childProcs {
+			_ = proc.Process.Kill()
+		}
+	}()
+
+	for i := 0; i < goMaxProcs; i++ {
 		/* #nosec G204 */
 		cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: I would prefer if you could merge this `exec.Command` setup with the one below. Maybe put it in a separate function?

File: prefork/prefork.go [hunk 4]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
-		if err := p.setTCPListenerFiles(addr); err != nil {
-			return err
+		if err = p.setTCPListenerFiles(addr); err != nil {
+			return
 		}
 
-		defer p.ln.Close()
+		defer func() {
+			if err == nil {
+				err = p.ln.Close()
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: Can you document why the listener only being closed if `!p.Reuseport`.

Comment (erikdubbelboer) [reply prefork/prefork.go]: And even if `err != nil` the listener still needs to be closed properly.

File: prefork/prefork.go [hunk 5]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
+			delete(childProcs, sig.pid)
+
+			log.Printf("one of the child prefork processes failed to complete, "+
+				"error: %v", sig.err)
+
+			if brokenProcs++; brokenProcs > p.RecoverThreshold {
+				log.Printf("child prefork processes exit too many times, "+
+					"which exceeds the value of RecoverThreshold(%d), "+
+					"exiting the master process.\n", brokenProcs)
+				err = errors.New("exceeding the value of RecoverThreshold")
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: This should be an exported package constant so users can compare against it.

File: prefork/prefork.go [hunk 6]
```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			cmd.ExtraFiles = p.files
+			childProcs[cmd.Process.Pid] = cmd
+			go func() {
+				sigCh <- procSig{cmd.Process.Pid, cmd.Run()}
+			}()
+		} else {
+			if completeProcs++; completeProcs == goMaxProcs {
+				break
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: I'm not following this logic. Is this for when the process exists without an error? Why wouldn't that use the same counter as `brokenProcs`?

File: prefork/prefork.go (hunk 1 above)
Comment (panjf2000) [reply prefork/prefork.go]: My oversight, fixed it.

File: prefork/prefork.go (hunk 2 above)
Comment (panjf2000) [reply prefork/prefork.go]: Got it.

File: prefork/prefork.go (hunk 3 above)
Comment (panjf2000) [reply prefork/prefork.go]: Good point.

File: prefork/prefork.go (hunk 5 above)
Comment (panjf2000) [reply prefork/prefork.go]: Done.

File: prefork/prefork.go (hunk 4 above)
Comment (panjf2000) [reply prefork/prefork.go]: > Can you document why the listener only being closed if !p.Reuseport.

You mean to add some comments on it?

Comment (erikdubbelboer) [reply prefork/prefork.go]: Yes please. I know the close was already there, but might as well add a comment if we're making changes here 😄

File: prefork/prefork.go (hunk 6 above)
Comment (panjf2000) [reply prefork/prefork.go]: > Why wouldn't that use the same counter as brokenProcs?

Sorry, I didn't get your point, two counter are for different purposes, or you were saying that we shouldn't exit the master process even if all child processes exit successfully?

File: prefork/prefork.go (hunk 4 above)
Comment (panjf2000) [reply prefork/prefork.go]: Done.

File: prefork/prefork.go (hunk 6 above)
Comment (panjf2000) [reply prefork/prefork.go]: You prefer to start over any exited child process despite of it is successful or unsuccessful?

Comment (erikdubbelboer) [reply prefork/prefork.go]: Yes. Does `os/exec` even return an error if the process exists with an error? Doesn't it only throw an error if there was some other cause like not being able to start the other process?

File: prefork/prefork.go [hunk 7]
```diff
@@ -110,30 +144,91 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
...
-		if err := p.setTCPListenerFiles(addr); err != nil {
-			return err
+		if err = p.setTCPListenerFiles(addr); err != nil {
+			return
 		}
 
-		defer p.ln.Close()
+		// defer for closing the net.Listener opened by setTCPListenerFiles.
+		defer func() {
+			e := p.ln.Close()
```
Comment (erikdubbelboer) [inline prefork/prefork.go]: Should the listener from reuseport not be closed?

**Comment by erikdubbelboer** [comment]
Isn't it better if RecoverThreshold has been reached for the master process to kill all the remaining children before returning?

File: server.go [hunk 8]
```diff
@@ -876,16 +876,21 @@ func (ctx *RequestCtx) FormFile(key string) (*multipart.FileHeader, error) {
 var ErrMissingFile = errors.New("there is no uploaded file associated with the given key")
 
 // SaveMultipartFile saves multipart file fh under the given filename path.
-func SaveMultipartFile(fh *multipart.FileHeader, path string) error {
-	f, err := fh.Open()
+func SaveMultipartFile(fh *multipart.FileHeader, path string) (err error) {
```
Comment (erikdubbelboer) [inline server.go]: I would prefer you make a separate pull request for these changes. I can them merge those into master and you can update this branch on top of master.

Comment (panjf2000) [reply server.go]: OK

**Comment by panjf2000** [comment]
> Isn't it better if RecoverThreshold has been reached for the master process to kill all the remaining children before returning?

What's the problem of killing them in `defer`?

File: prefork/prefork.go (hunk 7 above)
Comment (panjf2000) [reply prefork/prefork.go]: `p.ln` doesn't even get assigned an instance of `net.Listener` under reuseport mode.

Comment (panjf2000) [reply prefork/prefork.go]: Is the comment not good enough?

File: prefork/prefork.go (hunk 6 above)
Comment (panjf2000) [reply prefork/prefork.go]: Done.

**Comment by panjf2000** [comment]
Any progress？
@erikdubbelboer

**Comment by erikdubbelboer** [comment]
Thanks!

**Comment by savsgio** [comment]
Thanks!

**Pull request merged**